### PR TITLE
[Merged by Bors] - feat(MvPolynomial/Symmetric): the fundamental theorem

### PR DESCRIPTION
**Pull request by alreadydone**
Proves that in a polynomial ring in n variables over a commutative ring, the subalgebra of symmetric polynomials is freely generated by the first n elementary symmetric polynomials. This is expressed as an isomorphism with the symmetric subalgebra with the polynomials in n variables, `MvPolynomial.equiv_symmetricSubalgebra`.

Moves `Symmetric.lean` (renamed to `Defs.lean`), `NewtonIdentities.lean` to a new `Symmetric` folder where the new `FundamentalTheorem.lean` file also lives.

---

- [x] depends on: #6625
- [x] depends on: #7173
- [x] depends on: #7229
- [x] depends on: #16630

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/MvPolynomial/Symmetric/FundamentalTheorem.lean [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+    LinearOrderedCancelAddCommMonoid (Lex (σ →₀ β)) :=
+  { (inferInstance : LinearOrder (Lex (σ →₀ β))) with
+    add_le_add_left := fun a b h c => by
+      obtain rfl | ⟨i, h⟩ := h.eq_or_lt
+      · rfl
+      exact le_of_lt ⟨i, fun j hj => congr_arg (ofLex c j + ·) (h.1 j hj), add_lt_add_left h.2 _⟩
+    le_of_add_le_add_left := fun a b c h => by
+      obtain h | ⟨i, h⟩ := h.eq_or_lt
+      · exact (add_left_cancel h).le
+      exact le_of_lt ⟨i, fun j hj => (add_left_cancel <| h.1 j hj), lt_of_add_lt_add_left h.2⟩ }
```
Comment (eric-wieser) [inline Mathlib/RingTheory/MvPolynomial/Symmetric/FundamentalTheorem.lean:74]: This probably belongs in an earlier finsupp file; preferably with identical results for `Prod`, `Pi`, and `Dfinsupp` to keep the API for `Lex` in sync.

Comment (alreadydone) [reply Mathlib/RingTheory/MvPolynomial/Symmetric/FundamentalTheorem.lean:48]: #6625

File: Mathlib/RingTheory/MvPolynomial/Symmetric/FundamentalTheorem.lean [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+
+section CommRing
+
+variable (R) [Fintype σ] [CommRing R]
+
+-- shortcut instance necessary to avoid timeout
+instance {K} [CommRing K] : AddCommMonoid K := inferInstance
+
+/- May also holds for cancellative CommSemiring, not sure. -/
+lemma injective_esymmAlgHom_Fin (h : n ≤ m) :
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/MvPolynomial/Symmetric/FundamentalTheorem.lean:445]:
```suggestion
lemma injective_esymmAlgHom_fin (h : n ≤ m) :
```
and similar below

**Comment by alreadydone** [comment]
Thanks but note that generalized maxDegree API is almost ready at https://github.com/leanprover-community/mathlib4/compare/supDegree_LinearOrder

File: Mathlib/Data/Finsupp/Lex.lean [hunk 3]
```diff
@@ -78,9 +78,12 @@ instance Lex.partialOrder [PartialOrder N] : PartialOrder (Lex (α →₀ N)) :=
...
 /-- The linear order on `Finsupp`s obtained by the lexicographic ordering. -/
-instance Lex.linearOrder [LinearOrder N] : LinearOrder (Lex (α →₀ N)) :=
-  { @Lex.partialOrder α N _ _ _,  -- Porting note: Added types to avoid typeclass inference problem.
-    LinearOrder.lift' (toLex ∘ toDFinsupp ∘ ofLex) finsuppEquivDFinsupp.injective with }
+instance Lex.linearOrder [LinearOrder N] : LinearOrder (Lex (α →₀ N)) where
+  lt := (· < ·)
+  le := (· ≤ ·)
+  __ := @Lex.partialOrder α N _ _ _
+  -- Porting note: Added types to avoid typeclass inference problem.
+  __ := LinearOrder.lift' (toLex ∘ toDFinsupp ∘ ofLex) finsuppEquivDFinsupp.injective
```
Comment (alreadydone) [inline Mathlib/Data/Finsupp/Lex.lean:86]: Modified this instance for better defeq, which requires fixing (golfing) a counterexample

**Review by alreadydone — COMMENTED** [review]
PR is ready for review now using this [diff](https://github.com/leanprover-community/mathlib4/compare/supDegree_LinearOrder...FundThm_SymPoly#diff-3e39294da444cd659b1917031f8958fcb7c62cf4d2a920a5e4a3fc26e2a8609aR82) against the PR it depends on.

File: Mathlib/Data/Finsupp/Lex.lean (hunk 3 above)
Comment (eric-wieser) [reply Mathlib/Data/Finsupp/Lex.lean:86]: Can you split that into a standalone PR?

Comment (alreadydone) [reply Mathlib/Data/Finsupp/Lex.lean]: #7229

**Comment by alreadydone** [comment]
Because the diff from #7229 is large, maybe it's not a good time to review ...

**Comment by Ruben-VandeVelde** [comment]
I'll push a fix for the build errors in a bit

**Comment by alreadydone** [comment]
I was distracted by other projects and didn't finish the proposed change to the prerequisite PR #7173, so I'm not sure that fixing build errors at this time is particularly useful ...

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**